### ISSUE TITLE
firebase-functions on emulator behave differently on Windows vs on macOS

### ISSUE THREAD

**Issue by tkanq** [firebase/firebase-tools#4897]
### [REQUIRED] Environment info

**firebase-tools: 11.7.0**
**firebase-functions: 3.22.0**

**Platform: macOS 12.4 / Windows 10 Pro**
**Node.js: 14.19.3 (actually, here is another problem that if it's v16, my code doesn't work on Windows too...)***
**Java(for firebase-tools): 11.0.16**

### [REQUIRED] Test case

```
import functions from 'firebase-functions'
import https from 'https'
import axios from 'axios'
import {createHash} from 'crypto'

export default functions.region('asia-northeast1').https.onRequest(async (req, res) => {
  const sharedKey = "abcdefghijklmn"
  const randomKey = "opqrstuvwxyz"
  const sha256Hex = createHash("sha256").update(sharedKey+randomKey).digest("hex")
  const base64Encoded = Buffer.from(sha256Hex).toString('base64')

  await axios.post(
    `https://my.target.api.com`,
    {},
    {
      params: {
        auth_cd: base64Encoded,
        auth_key: randomKey,
        companyId: "myCompany"
      },
  // Below is just to avoid axios's error "Hostname/IP does not match certificate's
      httpsAgent: new https.Agent({
        rejectUnauthorized: false,
      }), 
      proxy: {
        protocol: "https",
        host: "my.squid.proxy.com",
        port: 8080,
        auth: {
          username: "username",
          password: "password"
        }
      }
    }
  )
  .then(result => {
    console.log(result)
  }).catch(error => {
    console.log(error);
  })

  res.status(200).end()
})
```
*some strings are replaced for security reason

### [REQUIRED] Steps to reproduce

I'm sorry that this is not reproducible on your environment because the external API we use above is not opened for everyone...

### [REQUIRED] Expected behavior

When we run the function above with local emulator, the OS on which emulator is running should not affect to the result of the functions.
But on Windows, axios returns expected response, and on macOS, it doesn't (I tried on two Mac: M1 and Intel, and neither doesn't work)

I also tried other http libraries: "got" and "superagent", but the result is same (expected result on windows and unexpected result on macOS) so I came here to ask if there is any clues, or if this is not related to firebase emulator.

### [REQUIRED] Actual behavior

[mac(unexpected).log](https://github.com/firebase/firebase-tools/files/9367126/mac.unexpected.log)
[windows(expected).log](https://github.com/firebase/firebase-tools/files/9367128/windows.expected.log)

*some strings are replaced for security reason

**Comment by taeold** [comment]
@tkanq Sorry to hear about your issues.

Based on my reading of the log output, I'm not very certain if this is related to Firebase Emulator 🤔 

I see error message like this

```
ERROR_LOGIN_AUTH_FAILED
```

which is coming from your internal backend service. Functions Emulator does monkey-patch the http module, but only to detect if your code is making external request (in this case, it does!) - it doesn't not modify your request in any way.

Just to confirm - can you not reproduce the error on a plain script (just write up some index.js file with axios request, run it)? Is this only happening when you are making the request to the backend from the Emulator using MacOS?

Relevant monkey-patching code:

https://github.com/firebase/firebase-tools/blob/master/src/emulator/functionsEmulatorRuntime.ts#L322-L390

I don't see anything suspicious, but I'll review it once more.

**Comment by tkanq** [comment]
@taeold 
Thank you for your help!

The answers to your questions:
> Just to confirm - can you not reproduce the error on a plain script (just write up some index.js file with axios request, run it)?

Now I turned my code into plain node.js code (I just deleted these lines below):
```
import functions from 'firebase-functions'
export default functions.region('asia-northeast1').https.onRequest(async (req, res) => {
})
```
and ran it on macOS.
The result is, it worked expectedly and got the same result with the one on Windows.
Can I say that it means the behavior of the codes changed when it is run on the function emulator?

Also, I read through that "Relevant monkey-patching code", and then tried one thing, which is to rebuild the source code with commenting out all the lines in initializeNetworkFiltering() function.
The result was same, so now I think it's clear that that monkey-patching code was not relevant to my problem.

Maybe my next action is to build my own api server and see the data or headers passed to it.

> Is this only happening when you are making the request to the backend from the Emulator using MacOS?

Yes, only on macOS this happens, and it doesn't happen on emulator on Windows.

**Comment by taeold** [comment]
@tkanq Thanks for sharing more details.

The situation sounds pretty frustrating - based on what you've shared here, it does sound like something in the Functions Emulator is interfering with the http request to your internal backend service. I'm surprised that disabling `initializeNetworkFiltering` didn't help - if anything that felt like the most suspicious.

I wonder if you can use something like https://stackoverflow.com/questions/41751235/how-to-log-all-axios-calls-from-one-place-in-code to intercept and log outgoing request to help us pinpoint how MacOS and Windows requests are different?

Thanks for your patience - looking forward to hearing back.

**Comment by tkanq** [comment]
@taeold 

I'm sorry that I couldn't respond to you for such a long time...

> I wonder if you can use something like https://stackoverflow.com/questions/41751235/how-to-log-all-axios-calls-from-one-place-in-code to intercept and log outgoing request to help us pinpoint how MacOS and Windows requests are different?
I did try what you suggested, but none of the logs seemed to be the clue to my problem.

So now I tried another experiment: I run same code on netlify functions emulator.
The result is completely same with firebase functions emulator, which means I can get expected result on Winodws, but not on Mac.

Now I think my problem is not specific to firebase-tools or firebase functions emulator, this issue should be stay closed and I should re-post this to other libraries (maybe axios or node.js?).

I appreciate your kind support, thank you.
